File: lib/src/main/java/com/asyncanimator/launcher/pending/PendingAnimation.java
package com.asyncanimator.launcher.pending;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.TimeInterpolator;
import android.animation.ValueAnimator;
import android.util.FloatProperty;

import com.asyncanimator.launcher.playback.AnimatorPlaybackController;
import com.asyncanimator.launcher.playback.PropertySetter;

import java.util.ArrayList;
import java.util.function.Consumer;

/**
 * PendingAnimation — 转场动画的"构建器"。
 *
 * <p>对应分析文档 §6.2。
 *
 * <p>关键设计：
 * <ul>
 *   <li>实现 {@link PropertySetter}（addFloat/setViewAlpha 等），让代码可以在"动画 vs no-op"之间无缝切换</li>
 *   <li>{@code add(Animator, TimeInterpolator, SpringProperty)} 把每个属性动画加到 {@link #mAnimHolders}</li>
 *   <li>{@code mProgressAnimator} 是辅助 ValueAnimator，挂 onEndListener / onFrameListener</li>
 *   <li>{@link #buildAnim()} 把 mProgressAnimator 合并到顶层 AnimatorSet，{@link #createPlaybackController()} 包装成 APC</li>
 * </ul>
 */
public class PendingAnimation implements PropertySetter {

    private final AnimatorSet mAnim = new AnimatorSet();
    private final ArrayList<AnimatorPlaybackController.Holder> mAnimHolders = new ArrayList<>();
    private final long mDuration;
    private ValueAnimator mProgressAnimator;
    private AnimatorPlaybackController mController;

    public boolean isAnimFinished = false;

    public PendingAnimation(long duration) {
        mDuration = duration > 0 ? duration : 0;
        mAnim.addListener(new AnimatorListenerAdapter() {
            @Override public void onAnimationStart(Animator a) { isAnimFinished = false; }
            @Override public void onAnimationEnd(Animator a)   { isAnimFinished = true; }
            @Override public void onAnimationCancel(Animator a) { isAnimFinished = true; }
        });
    }

    public PendingAnimation add(Animator anim) {
        anim.setDuration(mDuration);
        mAnim.playTogether(anim);
        addToHolders(anim);
        return this;
    }

    public PendingAnimation add(Animator anim, TimeInterpolator ip) {
        anim.setInterpolator(ip);
        return add(anim);
    }

    public PendingAnimation add(Animator anim, TimeInterpolator ip, Object springProperty) {
        anim.setInterpolator(ip);
        return add(anim);
    }

    public PendingAnimation addWithoutDuration(Animator anim) {
        mAnim.playTogether(anim);
        addToHolders(anim);
        return this;
    }

    public <T> PendingAnimation addFloat(T target, FloatProperty<T> property,
                                         float from, float to, TimeInterpolator ip) {
        ObjectAnimator oa = ObjectAnimator.ofFloat(target, property, from, to);
        oa.setInterpolator(ip);
        return add(oa.getAnimator());
    }

    @Override
    public <T> void setFloat(T target, FloatProperty<T> property, float value) {
        property.setValue(target, value);
    }

    public PendingAnimation addEndListener(Consumer<Boolean> consumer) {
        if (mProgressAnimator == null) mProgressAnimator = ValueAnimator.ofFloat(0f, 1f);
        mProgressAnimator.addListener(AnimatorListeners.forEndCallback(consumer));
        return this;
    }

    public PendingAnimation addOnFrameCallback(Runnable runnable) {
        if (mProgressAnimator == null) mProgressAnimator = ValueAnimator.ofFloat(0f, 1f);
        mProgressAnimator.addUpdateListener(animator -> runnable.run());
        return this;
    }

    public PendingAnimation addListener(Animator.AnimatorListener l) {
        mAnim.addListener(l);
        return this;
    }

    public AnimatorSet buildAnim() {
        if (mProgressAnimator != null) {
            addWithoutDuration(mProgressAnimator);
            mProgressAnimator = null;
        }
        if (mAnimHolders.isEmpty()) {
            addWithoutDuration(ValueAnimator.ofFloat(0f, 1f).setDuration(mDuration));
        }
        return mAnim;
    }

    public AnimatorPlaybackController createPlaybackController() {
        if (mController == null) {
            mController = new AnimatorPlaybackController(buildAnim(), mDuration, mAnimHolders);
        }
        return mController;
    }

    private void addToHolders(Animator anim) {
        AnimatorPlaybackController.addHoldersRecur(anim, mDuration, mAnimHolders);
    }

    // ──── 简化的 ObjectAnimator ────────────────────────────────

    public static class ObjectAnimator {
        private final Object target;
        private final FloatProperty property;
        private final float from;
        private final float to;
        private final ValueAnimator va = ValueAnimator.ofFloat(0f, 1f);

        public static <T> ObjectAnimator ofFloat(T target, FloatProperty<T> property,
                                                float from, float to) {
            return new ObjectAnimator(target, property, from, to);
        }

        /** public 以便 continuation 包的 TimeControllerObjectAnimator 子类可调用 super(...)。 */
        public ObjectAnimator(Object target, FloatProperty property, float from, float to) {            this.target = target;
            this.property = property;
            this.from = from;
            this.to = to;
        }

        public ObjectAnimator setInterpolator(TimeInterpolator ip) {
            va.setInterpolator(ip);
            return this;
        }

        public ObjectAnimator setDuration(long duration) {
            va.setDuration(duration);
            return this;
        }

        public ObjectAnimator setFloatValues(float... values) {
            va.setFloatValues(values);
            return this;
        }

        public Animator getAnimator() {
            // 内部 va 推进时把 fraction 映射到 from→to 再 setValue
            va.addUpdateListener(a -> {
                Float f = (Float) a.getAnimatedValue();
                if (f != null && property != null && target != null) {
                    float v = from + (to - from) * f;
                    property.setValue((Object) target, v);
                }
            });
            // 平台 Animator 的抽象方法比仿写件多（getStartDelay/setStartDelay/
            // setDuration/setInterpolator），全部委托给内部 va，保持原包装语义不变
            return new Animator() {
                @Override public void start() { va.start(); }
                @Override public void cancel() { va.cancel(); }
                @Override public void end() { va.end(); }
                @Override public boolean isRunning() { return va.isRunning(); }
                @Override public long getDuration() { return va.getDuration(); }
                @Override public Animator setDuration(long duration) { va.setDuration(duration); return this; }
                @Override public long getStartDelay() { return va.getStartDelay(); }
                @Override public void setStartDelay(long startDelay) { va.setStartDelay(startDelay); }
                @Override public void setInterpolator(TimeInterpolator ip) { va.setInterpolator(ip); }
            };
        }

        // 让 add() 可以直接接收 ObjectAnimator
        public long getDuration() { return va.getDuration(); }
        public TimeInterpolator getInterpolator() { return va.getInterpolator(); }
        public void setDurationOnly(long d) { va.setDuration(d); }

        // ──── 生命周期委托（TimeControllerObjectAnimator 依赖） ────────
        public void start() { va.start(); }
        public void cancel() { va.cancel(); }
        public void end() { va.end(); }
        public void pause() { va.pause(); }
        public boolean isRunning() { return va.isRunning(); }
        public void addListener(Animator.AnimatorListener l) { va.addListener(l); }
    }
}

File: lib/src/main/java/com/asyncanimator/launcher/pending/PendingAnimation.kt
package com.asyncanimator.launcher.pending

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.TimeInterpolator
import android.animation.ValueAnimator
import android.util.FloatProperty
import com.asyncanimator.launcher.playback.AnimatorPlaybackController
import com.asyncanimator.launcher.playback.PropertySetter
import java.util.function.Consumer

/**
 * PendingAnimation — 转场动画的"构建器"。
 *
 * 对应分析文档 §6.2。
 *
 * 关键设计：
 *
 *  - 实现 [PropertySetter]（addFloat/setViewAlpha 等），让代码可以在"动画 vs no-op"之间无缝切换
 *  - `add(Animator, TimeInterpolator, springProperty)` 把每个属性动画加到 [animHolders]
 *  - `progressAnimator` 是辅助 ValueAnimator，挂 onEndListener / onFrameListener
 *  - [buildAnim] 把 progressAnimator 合并到顶层 AnimatorSet，[createPlaybackController] 包装成 APC
 */
class PendingAnimation(duration: Long) : PropertySetter {

    private val anim = AnimatorSet()
    private val animHolders = ArrayList<AnimatorPlaybackController.Holder>()
    private val durationMs: Long = if (duration > 0) duration else 0
    private var progressAnimator: ValueAnimator? = null
    private var controller: AnimatorPlaybackController? = null

    var isAnimFinished = false

    init {
        anim.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationStart(a: Animator) { isAnimFinished = false }
            override fun onAnimationEnd(a: Animator) { isAnimFinished = true }
            override fun onAnimationCancel(a: Animator) { isAnimFinished = true }
        })
    }

    fun add(child: Animator): PendingAnimation {
        child.duration = durationMs
        anim.playTogether(child)
        addToHolders(child)
        return this
    }

    fun add(child: Animator, ip: TimeInterpolator?): PendingAnimation {
        child.interpolator = ip
        return add(child)
    }

    fun add(child: Animator, ip: TimeInterpolator?, springProperty: Any?): PendingAnimation {
        child.interpolator = ip
        return add(child)
    }

    fun addWithoutDuration(child: Animator): PendingAnimation {
        anim.playTogether(child)
        addToHolders(child)
        return this
    }

    fun <T> addFloat(target: T, property: FloatProperty<T>,
                     from: Float, to: Float, ip: TimeInterpolator?): PendingAnimation {
        val oa = ObjectAnimator.ofFloat(target, property, from, to)
        oa.setInterpolator(ip)
        return add(oa.buildAnimator())
    }

    override fun <T> setFloat(target: T, property: FloatProperty<T>, value: Float) {
        property.setValue(target, value)
    }

    fun addEndListener(consumer: Consumer<Boolean>?): PendingAnimation {
        if (progressAnimator == null) progressAnimator = ValueAnimator.ofFloat(0f, 1f)
        progressAnimator!!.addListener(AnimatorListeners.forEndCallback(consumer))
        return this
    }

    fun addOnFrameCallback(runnable: Runnable): PendingAnimation {
        if (progressAnimator == null) progressAnimator = ValueAnimator.ofFloat(0f, 1f)
        progressAnimator!!.addUpdateListener { runnable.run() }
        return this
    }

    fun addListener(l: Animator.AnimatorListener): PendingAnimation {
        anim.addListener(l)
        return this
    }

    fun buildAnim(): AnimatorSet {
        progressAnimator?.let {
            addWithoutDuration(it)
            progressAnimator = null
        }
        if (animHolders.isEmpty()) {
            addWithoutDuration(ValueAnimator.ofFloat(0f, 1f).setDuration(durationMs))
        }
        return anim
    }

    fun createPlaybackController(): AnimatorPlaybackController {
        if (controller == null) {
            controller = AnimatorPlaybackController(buildAnim(), durationMs, animHolders)
        }
        return controller!!
    }

    private fun addToHolders(child: Animator) {
        AnimatorPlaybackController.addHoldersRecur(child, durationMs, animHolders)
    }

    // ──── 简化的 ObjectAnimator ────────────────────────────────

    open class ObjectAnimator(
        private val target: Any?,
        private val property: FloatProperty<*>?,
        private val from: Float,
        private val to: Float
    ) {

        private val va: ValueAnimator = ValueAnimator.ofFloat(0f, 1f)

        open fun setInterpolator(ip: TimeInterpolator?): ObjectAnimator = apply {
            va.interpolator = ip
        }

        open fun setDuration(duration: Long): ObjectAnimator = apply {
            va.duration = duration
        }

        open fun setFloatValues(vararg values: Float): ObjectAnimator = apply {
            va.setFloatValues(*values)
        }

        /**
         * 产出一个平台 Animator：内部 va 推进时把 fraction 映射到 from→to 再 setValue。
         * 平台 Animator 的抽象方法比仿写件多（getStartDelay/setStartDelay/
         * setDuration/setInterpolator），全部委托给内部 va，保持原包装语义不变。
         */
        fun buildAnimator(): Animator {
            va.addUpdateListener { a ->
                val f = a.animatedValue as? Float
                if (f != null && property != null && target != null) {
                    val v = from + (to - from) * f
                    @Suppress("UNCHECKED_CAST")
                    (property as FloatProperty<Any?>).setValue(target, v)
                }
            }
            return object : Animator() {
                override fun start() = va.start()
                override fun cancel() = va.cancel()
                override fun end() = va.end()
                override fun isRunning(): Boolean = va.isRunning
                override fun getDuration(): Long = va.duration
                override fun setDuration(duration: Long): Animator = apply { va.duration = duration }
                override fun getStartDelay(): Long = va.startDelay
                override fun setStartDelay(startDelay: Long) { va.startDelay = startDelay }
                override fun setInterpolator(ip: TimeInterpolator?) { va.interpolator = ip }
            }
        }

        val duration: Long get() = va.duration

        // ──── 生命周期委托（TimeControllerObjectAnimator 依赖） ────────
        fun start() = va.start()
        fun cancel() = va.cancel()
        fun end() = va.end()
        fun pause() = va.pause()
        fun isRunning(): Boolean = va.isRunning
        fun addListener(l: Animator.AnimatorListener) = va.addListener(l)

        companion object {

            fun <T> ofFloat(target: T, property: FloatProperty<T>,
                            from: Float, to: Float): ObjectAnimator =
                ObjectAnimator(target, property, from, to)
        }
    }
}
